x/debug/debug.go

```diff
@@ -0,0 +1,197 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import (
+	"html/template"
+	"io"
+	"log"
+	"net/http"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/internal/introspection"
+)
+
+func NewHandler(d *yarpc.Dispatcher) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		w.Header().Set("Content-Type", "text/html; charset=utf-8")
+		render(w, req, d)
+	}
+}
+
+func render(w io.Writer, req *http.Request, dispatcher *yarpc.Dispatcher) {
+	data := struct {
+		Dispatchers     []introspection.DispatcherStatus
+		PackageVersions []introspection.PackageVersion
+	}{
+		Dispatchers: []introspection.DispatcherStatus{
+			dispatcher.Introspect(),
+		},
+		PackageVersions: yarpc.PackageVersions,
+	}
+
+	if err := pageTmpl.ExecuteTemplate(w, "Page", data); err != nil {
+		log.Printf("yarpc/debug: Failed executing template: %v", err)
+	}
+}
+
+var pageTmpl = template.Must(template.New("Page").Funcs(template.FuncMap{}).Parse(pageHTML))
+
+const pageHTML = `
+<html>
+	<head>
+	<title>/debug/yarpc</title>
+	<style type="text/css">
+		body {
+			font-family: "Courier New", Courier, monospace;
+		}
+		table {
+			color:#333333;
+			border-width: 1px;
+			border-color: #3A3A3A;
+			border-collapse: collapse;
+		}
+		table th {
+			border-width: 1px;
+			padding: 8px;
+			border-style: solid;
+			border-color: #3A3A3A;
+			background-color: #B3B3B3;
+		}
+		table td {
+			border-width: 1px;
+			padding: 8px;
+			border-style: solid;
+			border-color: #3A3A3A;
+			background-color: #ffffff;
+		}
+		header::after {
+			content: "";
+			clear: both;
+			display: table;
+		}
+		h1 {
+			width: 40%;
+			float: left;
+			margin: 0;
+		}
+		div.dependencies {
+			width: 60%;
+			float: left;
+			font-size: small;
+			text-align: right;
+		}
+	</style>
+	</head>
+	<body>
+
+<header>
+<h1>/debug/yarpc</h1>
+<div class="dependencies">
+	{{range .PackageVersions}}
+	<span>{{.Name}}={{.Version}}</span>
+	{{end}}
+</div>
+</header>
+
+{{range .Dispatchers}}
```

Comment:
You'll want to be able to provide a test template where you can basically just do this, and then compare the output written to the writer with a static string you provide.